frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BasePushBranchAction.java

```diff
@@ -28,17 +33,86 @@
  *  <li>{@link CommonDataKeys#PROJECT}</li>
  * </ul>
  */
-public abstract class BasePushBranchAction extends GitMacheteRepositoryReadyAction {
+public abstract class BasePushBranchAction extends GitMacheteRepositoryReadyAction
+    implements
+      IActionSourceAwareBranchNameProvider {
   private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
 
   protected final List<SyncToRemoteStatus.Relation> PUSH_ELIGIBLE_STATUSES = List.of(
       SyncToRemoteStatus.Relation.Ahead,
-      SyncToRemoteStatus.Relation.DivergedAndNewerThanRemote,
-      SyncToRemoteStatus.Relation.DivergedAndOlderThanRemote,
+      SyncToRemoteStatus.Relation.DivergedAndNewer,
+      SyncToRemoteStatus.Relation.DivergedAndOlder,
       SyncToRemoteStatus.Relation.Untracked);
 
+  @Override
   @UIEffect
-  protected void doPush(Project project, GitRepository preselectedRepository, String branchName) {
+  public void update(AnActionEvent anActionEvent) {
+    super.update(anActionEvent);
+
+    Presentation presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    Option<String> selectedBranchName = getNameOfBranchUnderAction(anActionEvent);
+
+    if (selectedBranchName.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Push disabled due to undefined branch name");
+      return;
+    }
+
+    Option<SyncToRemoteStatus> syncToRemoteStatus = getGitMacheteRepository(anActionEvent)
+        .flatMap(repo -> repo.getBranchByName(selectedBranchName.get()))
+        .map(branch -> branch.getSyncToRemoteStatus());
+
+    if (syncToRemoteStatus.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Push disabled due to undefined sync to remote status");
+      return;
+    }
+
+    SyncToRemoteStatus.Relation relation = syncToRemoteStatus.get().getRelation();
+    boolean isEnabled = PUSH_ELIGIBLE_STATUSES.contains(relation);
+
+    if (isEnabled) {
+      Option<Boolean> isSelectedEqualCurrent = getCurrentBranchNameIfManaged(anActionEvent)
```

Comment:
Maybe `getCurrentBranchNameIfManaged(anActionEvent).equals(Option.some(selectedBranchName.get()))`, or even `getCurrentBranchNameIfManaged(anActionEvent).equals(selectedBranchName)`? Not sure if 100% correct though/not insisting even if correct